sunscreen_math/src/metal_impl/mod.rs

```diff
@@ -0,0 +1,128 @@
+use std::mem::size_of;
+use std::ops::Deref;
+
+use lazy_static::lazy_static;
+use metal::{
+    Buffer, CommandQueue, Device, Library, MTLCommandBufferStatus, MTLResourceOptions, MTLSize,
+};
+
+mod ristrettovec;
+pub use ristrettovec::*;
+mod scalarvec;
+pub use scalarvec::*;
+
+// In build.rs, we compile 2 variants of the curve25519-dalek.metallib library:
+// test and release. In test, we #define the TEST macro, which exposes test kernels.
+// The release library does not feature these kernels.
+#[cfg(not(test))]
+const SHADERLIB: &[u8] = include_bytes!(concat!(
+    env!("OUT_DIR"),
+    "/curve25519-dalek.release.metallib"
+));
+
+#[cfg(test)]
+const SHADERLIB: &[u8] =
+    include_bytes!(concat!(env!("OUT_DIR"), "/curve25519-dalek.test.metallib"));
+
+pub struct Runtime {
+    device: Device,
+    lib: Library,
+    command_queue: CommandQueue,
+}
+
+/// TODO Remove
+unsafe impl Sync for Runtime {}
```

Comment:
More detail on why it is here now and what needs to be done to remove it.

Reply:
The real TODO is "is this even sound?" The real issue is that Runtime contains types that have mutable pointers in them, which makes them not automatically `impl Sync+Send`.

I don't know which (if any) Metal types can safely be shared between threads, but I needed to do this to make the POC work. I'll investigate this some more.

Command queues *do* appear to be [thread safe](https://developer.apple.com/library/archive/documentation/Miscellaneous/Conceptual/MetalProgrammingGuide/Cmd-Submiss/Cmd-Submiss.html#//apple_ref/doc/uid/TP40014221-CH3-SW14) so at least that part of the Runtime can be `Sync`.

Reply:
I wrapped the non-thread safe things in Mutex and command queue is thread safe. Thus, it is now sound to impl this.